b2/transferer.py

```diff
@@ -145,3 +230,79 @@ def as_info_dict(self):
             'contentSha1': self.content_sha1,
             'fileInfo': self.file_info,
         }
+
+
+class WriterThread(threading.Thread):
+    def __init__(self, file):
+        self.file = file
+        self.queue = queue.Queue()
+        super(WriterThread, self).__init__()
+
+    def run(self):
+        file = self.file
+        queue_get = self.queue.get
+        while 1:
+            shutdown, offset, data = queue_get()
+            if shutdown:
+                break
+            file.seek(offset)
+            file.write(data)
+
+    def __enter__(self):
+        self.start()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.queue.put((True, None, None))
+        self.join()
+
+
+class HashingDownloaderThread(threading.Thread):
+    def __init__(self, response, session, writer, hasher, range_):
+        self.response = response
+        self.session = session
+        self.writer = writer
+        self.hasher = hasher
+        assert range_[0] == 0, range_[0]
+        self.range_ = range_
+        super(HashingDownloaderThread, self).__init__()
+
+    def run(self):
+        writer_queue = self.writer.queue
+        stop = False
+        bytes_read = 0
+        hasher_update = self.hasher.update
+        first_offset = self.range_[0]
+        last_offset = self.range_[1]
+        for data in self.response.iter_content(chunk_size=BLOCK_SIZE):
+            if bytes_read + len(data) >= last_offset:
+                to_write = data[:last_offset - bytes_read]
+                stop = True
+            else:
+                to_write = data
+            writer_queue.put((False, first_offset + bytes_read, to_write))
+            hasher_update(to_write)
+            bytes_read += len(to_write)
+            if stop:
+                break
+        # since we got everything we need, close the socket and free the buffer
+        # to avoid a timeout exception during hashing and other trouble
+        self.response.close()
+
+
+class NonHashingDownloaderThread(threading.Thread):
+    def __init__(self, response, session, writer, range_):
+        self.session = session
+        self.url = response.request.url
+        self.writer = writer
+        self.range_ = range_
+        super(NonHashingDownloaderThread, self).__init__()
+
+    def run(self):
+        writer_queue_put = self.writer.queue.put
+        start_range = self.range_[0]
+        bytes_read = 0
+        with self.session.download_file_from_url(self.url, self.range_) as response:
```

Comment:
When downloading a range of a large file, the offset in the local file is different than the offset in the B2 file.  I think this class needs to be clear about whether the range is relative to the file stored in B2 or the local file on disk.

Reply:
If I understand correctly, then this is an implementation detail of the b2 server and nobody outside of backblaze knows that the large files are stored in separate chunks. For the purpose of the client, remote offset and local offset, are identical.

This class assumes that byte N in the cloud will be byte N in the file. Is the assumption incorrect?

Reply:
I think that assumption is incorrect when downloading a range of a file.  

Suppose `Transferer.download_file_from_url` is asked to download the range (50, 69) of a 100-byte file, and decides to download the file in two parts.  The result will be a 20-byte file.  When it fetches the second part, it will need to ask for range (60, 69) from the service, and write those 10 bytes at offset 10 in the local file.